File: python/studentdb/database/TrieNode.py
from typing import Dict, Optional

from studentdb.model.Stud import Stud


class TrieNode:
    """
    A node in the Trie data structure for storing students.

    Each node represents a character in a student's name and contains:

    - A map of child nodes for the next characters
    - An optional Student reference if this node represents the end of a name

    The Trie is case-insensitive, so all characters are stored in lowercase.
    """

    def __init__(self):
        self._children: Dict[str, 'TrieNode'] = {}
        self._student: Optional[Stud] = None

    def getChild(self, c: str) -> Optional['TrieNode']:
        """
        Returns the child node for the given character.

        Args:
            c: the character

        Returns:
            an Optional containing the child node if it exists, or None otherwise
        """
        return self._children.get(c.lower())

    def getOrCreateChild(self, c: str) -> 'TrieNode':
        """
        Creates and returns a child node for the given character if it doesn't exist.

        Args:
            c: the character

        Returns:
            the child node (existing or newly created)
        """
        lower_c = c.lower()
        if lower_c not in self._children:
            self._children[lower_c] = TrieNode()
        return self._children[lower_c]

    def getStudent(self) -> Optional[Stud]:
        """
        Returns the student stored at this node.

        Returns:
            the student if this node represents the end of a name
        """
        return self._student

    def setStudent(self, student: Optional[Stud]) -> None:
        """
        Stores a student at this node, marking this as the end of a name.

        Args:
            student: the student to store
        """
        self._student = student

    def hasStudent(self) -> bool:
        """
        Checks if this node has a student stored.

        Returns:
            true if a student is stored at this node, false otherwise
        """
        return self._student is not None

    def getChildren(self) -> Dict[str, 'TrieNode']:
        """
        Returns a map of all child nodes.

        Returns:
            the children map
        """
        return self._children

File: java/app/src/main/java/studentdb/database/TrieNode.java
package studentdb.database;

import java.util.HashMap;
import java.util.Map;
import java.util.Optional;
import studentdb.model.Stud;

/**
 * A node in the Trie data structure for storing students.
 *
 * <p>Each node represents a character in a student's n and contains:
 *
 * <ul>
 *   <li>A map of child nodes for the next characters
 *   <li>An optional Student reference if this node represents the end of a n
 * </ul>
 *
 * <p>The Trie is case-insensitive, so all characters are stored in lowercase.
 */
class TrieNode {

  private final Map<Character, TrieNode> children;
  private Stud student;

  TrieNode() {
    this.children = new HashMap<>();
    this.student = null;
  }

  /**
   * Returns the child node for the given character.
   *
   * @param c the character
   * @return an Optional containing the child node if it exists, or empty otherwise
   */
  Optional<TrieNode> getChild(char c) {
    return Optional.ofNullable(children.get(Character.toLowerCase(c)));
  }

  /**
   * Creates and returns a child node for the given character if it doesn't exist.
   *
   * @param c the character
   * @return the child node (existing or newly created)
   */
  TrieNode getOrCreateChild(char c) {
    return children.computeIfAbsent(Character.toLowerCase(c), k -> new TrieNode());
  }

  /**
   * Returns the student stored at this node.
   *
   * @return an Optional containing the student if this node represents the end of a n, or empty
   *     otherwise
   */
  Stud getStudent() {
    if (student == null) {
      throw new NullPointerException("Asked for student, but not available");
    }
    return student;
  }

  /**
   * Stores a student at this node, marking this as the end of a n.
   *
   * @param student the student to store
   */
  void setStudent(Stud student) {
    this.student = student;
  }

  /**
   * Checks if this node has a student stored.
   *
   * @return true if a student is stored at this node, false otherwise
   */
  boolean hasStudent() {
    return student != null;
  }

  /**
   * Returns a map of all child nodes.
   *
   * @return the children map
   */
  Map<Character, TrieNode> getChildren() {
    return children;
  }
}
